cli/src/tests/query_test.rs

```diff
@@ -0,0 +1,689 @@
+use super::helpers::allocations;
+use super::helpers::fixtures::get_language;
+use tree_sitter::{Node, Parser, Query, QueryCursor, QueryError, QueryMatch};
+
+#[test]
+fn test_query_errors_on_invalid_syntax() {
+    allocations::record(|| {
+        let language = get_language("javascript");
+
+        assert!(Query::new(language, "(if_statement)").is_ok());
+        assert!(Query::new(language, "(if_statement condition:(identifier))").is_ok());
+
+        // Mismatched parens
+        assert_eq!(
+            Query::new(language, "(if_statement"),
+            Err(QueryError::Syntax(13))
+        );
+        assert_eq!(
+            Query::new(language, "(if_statement))"),
+            Err(QueryError::Syntax(14))
+        );
+
+        // Return an error at the *beginning* of a bare identifier not followed a colon.
+        // If there's a colon but no pattern, return an error at the end of the colon.
+        assert_eq!(
+            Query::new(language, "(if_statement identifier)"),
+            Err(QueryError::Syntax(14))
+        );
+        assert_eq!(
+            Query::new(language, "(if_statement condition:)"),
+            Err(QueryError::Syntax(24))
+        );
+
+        // Return an error at the beginning of an unterminated string.
+        assert_eq!(
+            Query::new(language, r#"(identifier) "h "#),
+            Err(QueryError::Syntax(13))
+        );
+
+        assert_eq!(
+            Query::new(language, r#"((identifier) ()"#),
+            Err(QueryError::Syntax(16))
+        );
+        assert_eq!(
+            Query::new(language, r#"((identifier) @x (eq? @x a"#),
+            Err(QueryError::Syntax(26))
+        );
+    });
+}
```

Comment:
The `TSQuery` API is mainly tested through these Rust unit tests. There are also some additional tests in the JavaScript-based WASM test suite.